coordinator/src/funding_fee.rs

```diff
@@ -0,0 +1,277 @@
+
+        Self {
+            rate,
+            start_date,
+            end_date,
+        }
+    }
+
+    pub fn rate(&self) -> Decimal {
+        self.rate
+    }
+
+    pub fn start_date(&self) -> OffsetDateTime {
+        self.start_date
+    }
+
+    pub fn end_date(&self) -> OffsetDateTime {
+        self.end_date
+    }
+}
+
+/// A record that a funding fee is owed between the coordinator and a trader.
+#[derive(Clone, Debug)]
+pub struct FundingFeeEvent {
+    pub id: i32,
+    /// A positive amount indicates that the trader pays the coordinator; a negative amount
+    /// indicates that the coordinator pays the trader.
+    pub amount: SignedAmount,
+    pub trader_pubkey: PublicKey,
+    pub position_id: i32,
+    pub due_date: OffsetDateTime,
+    pub paid_date: Option<OffsetDateTime>,
+}
+
+#[derive(Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub enum IndexPriceSource {
+    Bitmex,
+    Fake,
+}
+
+pub async fn generate_funding_fee_events_periodically(
+    scheduler: &JobScheduler,
+    pool: Pool<ConnectionManager<PgConnection>>,
+    schedule: String,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    scheduler
+        .add(tokio_cron_scheduler::Job::new(
+            schedule.as_str(),
+            move |_, _| {
+                if let Err(e) = generate_funding_fee_events(&pool, index_price_source) {
+                    tracing::error!("Failed to generate funding fee events: {e:#}");
+                }
+            },
+        )?)
+        .await?;
+
+    scheduler.start().await?;
+
+    Ok(())
+}
+
+/// Generate [`FundingFeeEvent`]s for all active positions.
+///
+/// When called, a [`FundingFeeEvent`] will be generated for an active position if:
+///
+/// - We can get a [`FundingRate`] that is at most 1 hour old from the DB.
+/// - We can get a BitMEX index price for the `end_date` of the [`FundingRate`].
+/// - There is no other [`FundingFeeEvent`] in the DB with the same `position_id` and `end_date`.
+/// - The position was created _before_ the `end_date` of the [`FundingRate`].
+fn generate_funding_fee_events(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    let mut conn = pool.get()?;
+
+    let funding_rate = db::funding_rates::get_funding_rate_charged_in_the_last_hour(&mut conn)?;
+
+    let funding_rate = match funding_rate {
+        Some(funding_rate) => funding_rate,
+        None => {
+            tracing::debug!("No current funding rate for this hour");
+            return Ok(());
+        }
+    };
+
+    // TODO: Funding rates should be specific to contract symbols.
+    let contract_symbol = ContractSymbol::BtcUsd;
+
+    let current_index_price = match index_price_source {
+        IndexPriceSource::Bitmex => block_in_place(move || {
+            let current_index_price =
+                get_bitmex_index_price(&contract_symbol, funding_rate.end_date)?;
```

Comment:
I was thinking about this too, although my motivation would have been to make it easier to write e2e tests for this feature.

But I remember that we wanted to ensure that we get a price here, and polling means that we avoid timing issues with the separate task you propose.

We obviously can't really ensure that we will always have a price, so we actually need to be able to retry this task until we get the information we need. I suppose that means that we can safely move the index-price-fetching into a separate task.

Reply:
Yes, let's do it.

Reply:
I'm actually not sure we want this.

It would only aid testing if it was controlled from the outside, but that's another annoying thing to manage from the outside, which makes the coordinator less self-sufficient.

To your point, I'm not yet convinced we will need to send the index price over to the app. I propose to add it once we know what we want to do with it for sure.

> We obviously can't really ensure that we will always have a price, so we actually need to be able to retry this task until we get the information we need.

W.r.t. this thing I mentioned, I've added some logic to retry a few times if we encounter errors.